Fixed some issues with inotify and multiple events at the same time.

adamharrison opened the pull request:
Seems to be working now.

jgmdev commented:
Fixed the aka git clone infinite loop bug details on the patch below which can be applied using `git am 0001-dirmonitor-fix-infinite-loop-on-dir-add.patch`:

0001-dirmonitor-fix-infinite-loop-on-dir-add.patch
```patch
From 59e10dff2113bffba7c8df5bc8f43d6cc0fd6194 Mon Sep 17 00:00:00 2001
From: jgmdev <jgmdev@gmail.com>
Date: Tue, 8 Mar 2022 03:59:26 -0400
Subject: [PATCH 1/1] dirmonitor: fix infinite loop on dir add

The issue was first discovered when doing a git clone on a project
directory that was opened on lite-xl using new dirmonitor changes.
This caused the treeview to go into an infinite loop because the
is_dirty flag was never set by the dirmonitor for the reason explained
below.

This was actually a major pain to troubleshoot, the function
`project_subdir_bounds(dir, filename)` wasn't returning appropiate
values when a dir/file wasn't on the internal project files list, and
the code that called it was doing aritmethic operations with the
returned nil value... but for a strange reason no error was been thrown by
lite-xl when trying to perform a `nil + number` sum (maybe because the code
is running inside of a callback function been called from dirmonitor c code
that isn't doing any error handling?) anyways, it should have thrown the
following error but it didn't making it a pain to debug (at least to me):

`stdin:1: attempt to perform arithmetic on a nil value`

Besides that, the code that was sending the new dir for it to be added
to list and content refreshed was prepending a PATHSEP to the new dir name
which caused the directory to be shown as empty even when it had content
in it.

Time to fix: 4 hours, end of report :)
---
 data/core/init.lua | 12 +++++++-----
 1 file changed, 7 insertions(+), 5 deletions(-)

diff --git a/data/core/init.lua b/data/core/init.lua
index 116631d..ae12abb 100644
--- a/data/core/init.lua
+++ b/data/core/init.lua
@@ -189,10 +189,9 @@ end
 local function project_subdir_bounds(dir, filename)
   local index, n = 0, #dir.files
   for i, file in ipairs(dir.files) do
-    local file = dir.files[i]
     if file.filename == filename then
       index, n = i, #dir.files - i
-      for j = 1, #dir.files - i do
+      for j = 1, n do
         if not common.path_belongs_to(dir.files[i + j].filename, filename) then
           n = j - 1
           break
@@ -201,6 +200,7 @@ local function project_subdir_bounds(dir, filename)
       return index, n, file
     end
   end
+  return #dir.files, 0, nil
 end


@@ -273,9 +273,11 @@ local function refresh_directory(topdir, target, expanded)
         old_idx = old_idx + 1
       end
     end
-    for i, v in ipairs(new_directories) do
-      topdir.watch:watch(topdir.name .. PATHSEP .. (target ~= "" and (target .. PATHSEP) or "") .. v.filename)
-      if refresh_directory(topdir, target .. PATHSEP .. v.filename) then
+    for _, v in ipairs(new_directories) do
+      local watch_root = (topdir.name ~= "" and topdir.name .. PATHSEP or "")
+      local target_path = (target ~= "" and (target .. PATHSEP) or "") .. v.filename
+      topdir.watch:watch(watch_root .. target_path)
+      if refresh_directory(topdir, target_path) then
         change = true
       end
     end
--
2.35.1
```

adamharrison commented:
Yeah; good call. To just reiterate what we talked about on discord; there are still some issues with the code when dealing with many subdirectories being added at once with those changes; but yeah, there's a problem here with errors not bubbling up correctly as well. I've changed the C api to lua_call from lua_pcall, which should bubble up there error. I've also cleaned up the project_subdir_bounds function like you mentioned, as you pointed out there was a bit of redundancy in there.

adamharrison commented:
OK, both me and jgmdev have reviewed; I'm going to merge this in now, as this is basically a bugfix.
